File: oscexchange/src/main/java/ac/at/fhstp/digitech/oscexchange/LiveOSCExchange.java
package ac.at.fhstp.digitech.oscexchange;

import com.illposed.osc.MessageSelector;
import com.illposed.osc.OSCMessage;
import com.illposed.osc.OSCMessageListener;
import com.illposed.osc.OSCSerializeException;
import com.illposed.osc.messageselector.OSCPatternAddressMessageSelector;
import com.illposed.osc.transport.udp.OSCPortIn;
import com.illposed.osc.transport.udp.OSCPortOut;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.Optional;

import ac.at.fhstp.digitech.oscexchange.errors.OSCError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCMessageError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCParsingError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCPortClosingError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCPortOpeningError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCPortUseError;
import ac.at.fhstp.digitech.oscexchange.errors.OSCValidationError;

class LiveOSCExchange {

    static void start(OSCExchange exchange, OSCDevicePair devicePair) {
        Optional<OSCPortOut> outPort = tryGenerateOutPort(devicePair.remote);
        Optional<OSCPortIn> inPort = tryGenerateInPort(devicePair.local);

        if (!outPort.isPresent())
            fail(exchange, new OSCPortOpeningError(OSCPort.Out));
        else if (!inPort.isPresent())
            fail(exchange, new OSCPortOpeningError(OSCPort.In));
        else
            run(new LiveOSCExchange(exchange, outPort.get(), inPort.get()));
    }

    private static void fail(OSCExchange exchange, OSCError error) {
        if (exchange.errorListener != null)
            exchange.errorListener.handle(error);
    }

    private static void complete(OSCExchange exchange) {
        exchange.completeListener.handle();
    }

    private static Optional<OSCPortOut> tryGenerateOutPort(InetSocketAddress address) {
        try {
            return Optional.of(new OSCPortOut(address));
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    private static Optional<OSCPortIn> tryGenerateInPort(InetSocketAddress address) {
        try {
            OSCPortIn port = new OSCPortIn(address);
            return Optional.of(port);
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    private static void run(LiveOSCExchange live) {
        handleNextRequest(live);
    }

    private static boolean hasRequest(LiveOSCExchange live) {
        return live.requestIndex < OSCExchange.getRequestCount(live.exchange);
    }

    private static void complete(LiveOSCExchange live) {
        try {
            live.inPort.close();
        } catch (Exception e) {
            fail(live.exchange, new OSCPortClosingError(OSCPort.In));
            return;
        }

        try {
            live.outPort.close();
        } catch (Exception e) {
            fail(live.exchange, new OSCPortClosingError(OSCPort.Out));
            return;
        }

        complete(live.exchange);
    }

    private static void handleNextRequest(LiveOSCExchange live) {
        if (hasRequest(live)) {
            OSCRequest request = OSCExchange.getRequest(live.exchange, live.requestIndex);
            live.requestIndex++;
            handleRequest(live, request);
        } else
            complete(live);
    }

    private static void handleRequest(LiveOSCExchange live, OSCRequest request) {
        if (request instanceof OSCSendRequest)
            handleSendRequest(live, (OSCSendRequest) request);
        else if (request instanceof OSCRawReceiveRequest)
            handleRawReceiveRequest(live, (OSCRawReceiveRequest) request);
        else if (request instanceof OSCParsedReceiveRequest)
            handleParsedReceiveRequest(live, (OSCParsedReceiveRequest<?>) request);
    }

    private static void handleSendRequest(LiveOSCExchange live, OSCSendRequest request) {
        try {
            OSCMessage message = OSCSendRequest.generateMessage(request);
            live.outPort.send(message);
            handleNextRequest(live);
        } catch (IOException e) {
            fail(live.exchange, new OSCPortUseError(OSCPort.Out));
        } catch (OSCSerializeException e) {
            fail(live.exchange, new OSCMessageError(request.args));
        }
    }

    private static void handleReceiveRequest(LiveOSCExchange live, OSCReceiveRequest request, OSCRawReceiveListener onValidated) {
        MessageSelector selector = new OSCPatternAddressMessageSelector(request.address.value);
        ForwardRef<OSCMessageListener> listener = new ForwardRef<>();

        listener.value = event -> {
            live.inPort.getDispatcher().removeListener(selector, listener.value);
            live.inPort.stopListening();

            OSCArgs receivedArgs = OSCArgs.multiple(event.getMessage().getArguments());

            if (request.validator != null)
                if (request.validator.isValid(receivedArgs)) {
                    onValidated.handle(receivedArgs);
                    handleNextRequest(live);
                } else
                    fail(live.exchange, new OSCValidationError(receivedArgs));
            else
                handleNextRequest(live);
        };

        live.inPort.getDispatcher().addListener(selector, listener.value);
        live.inPort.startListening();
    }

    private static void handleRawReceiveRequest(LiveOSCExchange live, OSCRawReceiveRequest request) {
        handleReceiveRequest(live, request, request.listener);
    }

    private static <T> void handleParsedReceiveRequest(LiveOSCExchange live, OSCParsedReceiveRequest<T> request) {
        handleReceiveRequest(live, request, args -> {
            Optional<T> parsed = request.parser.parse(args);

            if (parsed.isPresent())
                request.listener.handle(parsed.get());
            else
                fail(live.exchange, new OSCParsingError(args));
        });
    }


    private final OSCExchange exchange;
    private final OSCPortOut outPort;
    private final OSCPortIn inPort;
    private int requestIndex = 0;


    private LiveOSCExchange(OSCExchange exchange, OSCPortOut outPort, OSCPortIn inPort) {
        this.exchange = exchange;
        this.outPort = outPort;
        this.inPort = inPort;
    }

}

File: oscexchange/src/main/java/ac/at/fhstp/digitech/oscexchange/LiveOSCExchange.kt
package ac.at.fhstp.digitech.oscexchange

import ac.at.fhstp.digitech.oscexchange.errors.*
import com.illposed.osc.*
import com.illposed.osc.messageselector.OSCPatternAddressMessageSelector
import com.illposed.osc.transport.udp.OSCPortIn
import com.illposed.osc.transport.udp.OSCPortOut
import java.io.IOException
import java.net.InetSocketAddress
import java.util.*

internal class LiveOSCExchange private constructor(
    private val exchange: OSCExchange,
    private val outPort: OSCPortOut,
    private val inPort: OSCPortIn
) {
    private var requestIndex = 0

    companion object {

        fun start(exchange: OSCExchange, devicePair: OSCDevicePair) {
            val outPort = tryGenerateOutPort(devicePair.remote)
            val inPort = tryGenerateInPort(devicePair.local)

            if (!outPort.isPresent)
                fail(
                    exchange, OSCPortOpeningError(null, OSCPort.Out)
                ) else if (!inPort.isPresent)
                fail(
                    exchange, OSCPortOpeningError(null, OSCPort.In)
                ) else run(
                LiveOSCExchange(
                    exchange,
                    outPort.get(),
                    inPort.get()
                )
            )
        }

        private fun fail(exchange: OSCExchange, error: OSCError) {
            if (exchange.errorListener != null)
                exchange.errorListener.handle(error)
        }

        private fun complete(exchange: OSCExchange) {
            if (exchange.completeListener != null)
                exchange.completeListener.handle()
        }

        private fun tryGenerateOutPort(address: InetSocketAddress?): Optional<OSCPortOut> {
            return try {
                Optional.of(OSCPortOut(address))
            } catch (e: Exception) {
                Optional.empty()
            }
        }

        private fun tryGenerateInPort(address: InetSocketAddress?): Optional<OSCPortIn> {
            return try {
                val port = OSCPortIn(address)
                Optional.of(port)
            } catch (e: Exception) {
                Optional.empty()
            }
        }

        private fun run(live: LiveOSCExchange) {
            handleNextRequest(live)
        }

        private fun hasRequest(live: LiveOSCExchange): Boolean {
            return live.requestIndex < OSCExchange.getRequestCount(
                live.exchange
            )
        }

        private fun complete(live: LiveOSCExchange) {
            try {
                live.inPort.close()
            } catch (e: Exception) {
                fail(live.exchange, OSCPortClosingError(e, OSCPort.In))
                return
            }
            try {
                live.outPort.close()
            } catch (e: Exception) {
                fail(live.exchange, OSCPortClosingError(e, OSCPort.Out))
                return
            }
            complete(live.exchange)
        }

        private fun handleNextRequest(live: LiveOSCExchange) {
            if (hasRequest(live)) {
                val request: OSCRequest = OSCExchange.getRequest(
                    live.exchange,
                    live.requestIndex
                )
                live.requestIndex++
                handleRequest(live, request)
            } else complete(live)
        }

        private fun handleRequest(live: LiveOSCExchange, request: OSCRequest) {
            when (request) {
                is OSCSendRequest ->
                    handleSendRequest(live, request)
                is OSCRawReceiveRequest ->
                    handleRawReceiveRequest(live, request)
                is OSCParsedReceiveRequest<*> ->
                    handleParsedReceiveRequest(live, request)
            }
        }

        private fun handleSendRequest(
            live: LiveOSCExchange,
            request: OSCSendRequest
        ) {
            try {
                val message: OSCMessage =
                    OSCSendRequest.generateMessage(request)
                live.outPort.send(message)
                handleNextRequest(live)
            } catch (e: IOException) {
                fail(live.exchange, OSCPortUseError(e, OSCPort.Out))
            } catch (e: OSCSerializeException) {
                fail(live.exchange, OSCMessageError(e, request.args))
            }
        }

        private fun handleReceiveRequest(
            live: LiveOSCExchange,
            request: OSCReceiveRequest,
            onValidated: OSCRawReceiveListener
        ) {
            val selector: MessageSelector =
                OSCPatternAddressMessageSelector(request.address.value)
            val listener = ForwardRef<OSCMessageListener>()
            listener.value = OSCMessageListener { event: OSCMessageEvent ->
                live.inPort.dispatcher.removeListener(selector, listener.value)
                live.inPort.stopListening()
                val receivedArgs: OSCArgs =
                    OSCArgs.list(event.message.arguments)
                if (request.validator != null) if (request.validator.isValid(
                        receivedArgs
                    )
                ) {
                    onValidated.handle(receivedArgs)
                    handleNextRequest(live)
                } else fail(
                    live.exchange,
                    OSCValidationError(null, receivedArgs)
                ) else handleNextRequest(live)
            }
            live.inPort.dispatcher.addListener(selector, listener.value)
            live.inPort.startListening()
        }

        private fun handleRawReceiveRequest(
            live: LiveOSCExchange,
            request: OSCRawReceiveRequest
        ) {
            handleReceiveRequest(
                live,
                request
            ) { args: OSCArgs ->
                if (request.listener != null)
                    request.listener.handle(args)
            }
        }

        private fun <T> handleParsedReceiveRequest(
            live: LiveOSCExchange,
            request: OSCParsedReceiveRequest<T>
        ) {
            handleReceiveRequest(live, request) { args: OSCArgs ->
                val parsed = request.parser.parse(args)
                if (parsed.isPresent) {
                    if (request.listener != null)
                        request.listener.handle(parsed.get())
                } else
                    fail(live.exchange, OSCParsingError(null, args))
            }
        }
    }
}